File: lib-segment/src/main/java/websiteschema/mpsegment/pinyin/WordToPinyinClassifier.java
package websiteschema.mpsegment.pinyin;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import websiteschema.mpsegment.core.SegmentResult;
import websiteschema.mpsegment.hmm.Node;
import websiteschema.mpsegment.hmm.ObserveListException;

import java.util.ArrayList;
import java.util.List;

public class WordToPinyinClassifier {

    WordToPinyinModel model;
    final static Log l = LogFactory.getLog("segment");

    public void setModel(WordToPinyinModel model) {
        this.model = model;
    }

    public void classify(SegmentResult result) {
        try {
            String originalStr = result.toOriginalString();
            List<String> pinyinList = classify(originalStr);
            int pos = 0;
            for (int i = 0; i < result.length(); i++) {
                int wordLength = result.getWord(i).length();
                String pinyin = join(pinyinList.subList(pos, pos += wordLength), "'");
                result.setPinyin(i, pinyin);
            }
        } catch (ObserveListException ex) {
            l.error(ex);
        }
    }

    private String join(List<String> list, String sep) {
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < list.size(); i++) {
            String str = list.get(i);
            stringBuilder.append(str);
            if (null != sep && i < list.size() - 1) {
                stringBuilder.append(sep);
            }
        }
        return stringBuilder.toString();
    }

    public List<String> classify(String o) throws ObserveListException {
        List<String> observeList = new ArrayList<String>(o.length());
        for (int i = 0; i < o.length(); i++) {
            observeList.add(String.valueOf(o.charAt(i)));
        }
        return classify(observeList);
    }

    private List<String> classify(List<String> o) throws ObserveListException {
        assert (null != o && o.size() > 0);
        List<Section> sections = findSectionsByUnknownCharacter(o);
        return classifySectionList(sections);
    }

    private List<String> classifySectionList(List<Section> sections) throws ObserveListException {
        List<String> result = new ArrayList<String>();
        for (Section section : sections) {
            if (section.hasKnownCharacters()) {
                List<String> observeCharacters = section.characters;
                result.addAll(convert(classifyOberveList(observeCharacters)));
            }
            if (section.hasUnknwonCharacter()) {
                result.add(section.unknownChar);
            }
        }
        return result;
    }

    private List<Node> classifyOberveList(List<String> observeCharacters) throws ObserveListException {
        return model.getViterbi().calculateWithLog(observeCharacters);
    }

    private List<Section> findSectionsByUnknownCharacter(List<String> o) {
        int lastSectorPos = 0;
        List<Section> sections = new ArrayList<Section>();
        for (int i = 0; i < o.size(); i++) {
            String ch = o.get(i);
            boolean knownObserveNode = model.knwonObserve(ch);
            if (!knownObserveNode) {
                Section section = new Section(o, lastSectorPos, i);
                sections.add(section);
                lastSectorPos = i + 1;
            }
        }
        Section sec = new Section(o, lastSectorPos);
        sections.add(sec);
        return sections;
    }

    private List<String> convert(List<Node> nodeList) {
        List<String> result = new ArrayList<String>(nodeList.size());
        for (Node node : nodeList) {
            result.add(node.getName());
        }
        return result;
    }

    class Section {
        List<String> characters;
        String unknownChar;

        Section(List<String> o, int start, int end) {
            if (end > start) {
                characters = new ArrayList<String>();
                for (int i = 0; i < end - start; i++) {
                    characters.add(o.get(i + start));
                }
                unknownChar = o.get(end);
            } else if (end == start) {
                unknownChar = o.get(end);
            }
        }

        Section(List<String> o, int start) {
            if (o.size() > start) {
                characters = new ArrayList<String>();
                for (int i = start; i < o.size(); i++) {
                    characters.add(o.get(i));
                }
            }
        }

        boolean hasKnownCharacters() {
            return null != characters;
        }

        boolean hasUnknwonCharacter() {
            return null != unknownChar;
        }
    }

}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/pinyin/WordToPinyinClassifier.scala
package websiteschema.mpsegment.pinyin

import websiteschema.mpsegment.core.SegmentResult
import websiteschema.mpsegment.hmm.{HmmClassifier, HmmModel}
import websiteschema.mpsegment.util.CharCheckUtil

import io.Source
import websiteschema.mpsegment.util.FileUtil._
import scala.Some

class WordToPinyinClassifier {

  private val classifier = new HmmClassifier()

  def setModel(model: HmmModel) {
    classifier setModel model
  }

  def getModel() = classifier.model

  def classify(result: SegmentResult) {
    try {
      val originalStr = result.toOriginalString()
      val pinyinList = classify(originalStr)
      var pos = 0
      for (i <- 0 until result.length) {
        val wordLength = result.getWord(i).length()
        val pinyin = join(subList(pinyinList, pos, pos + wordLength), "'", result.getWord(i))
        pos += wordLength
        result.setPinyin(i, pinyin)
      }
    } catch {
      case ex: Throwable =>
        System.err.println(ex)
    }
  }

  private def subList(list: List[String], start: Int, end: Int): List[String] = {
    list.drop(start).dropRight(list.size - end)
  }

  private def join(list: List[String], sep: String, word: String): String = {
    val stringBuilder = new StringBuilder()
    for (i <- 0 until list.size) {
      val str = list(i)
      stringBuilder.append(str)
      if (null != sep && i < list.size - 1 && CharCheckUtil.isChinese(word.charAt(i + 1))) {
        stringBuilder.append(sep)
      }
    }
    stringBuilder.toString()
  }

  def classify(o: String) = {
    var observeList = List[String]()
    for (i <- 0 until o.length) {
      observeList = observeList ++ List[String](o.charAt(i).toString)
    }
    classifier.classify(observeList)
  }

  def loadDictionary(dictFile : String) {
    val source = Source.fromInputStream(getResourceAsStream(dictFile))
    for (line <- source.getLines()) {
      val regex = "([^ ]+) *= *([^ ]+)".r
      regex findFirstIn line match {
        case Some(regex(han, pinyin)) => getModel().add(han, pinyin)
        case None =>
      }
    }
  }
}

